CMakeLists.txt

```diff
@@ -2,93 +2,279 @@
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
-if ( NOT "${SPICY_IN_TREE_BUILD}" )
-    set(ZEEK_SPICY_PLUGIN_BUILD_DIR ${CMAKE_BINARY_DIR}/plugin)
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${ZEEK_SPICY_PLUGIN_BUILD_DIR}/bin)
-
-    if ( NOT "${CMAKE_BUILD_TYPE}" )
-        # We follow Zeek's build mode by default.
-        if ( ZEEK_DEBUG_BUILD )
-            set(CMAKE_BUILD_TYPE "Debug" CACHE STRING "" FORCE)
-        else ()
-            set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "" FORCE)
-        endif ()
+# CMake uses -O2 by default with RelWithDebInfo.
+string(REPLACE "-O2" "-O3" CMAKE_CXX_FLAGS_RELWITHDEBINFO ${CMAKE_CXX_FLAGS_RELWITHDEBINFO})
+
+include_directories(BEFORE ${PROJECT_SOURCE_DIR}/include)
+include_directories(BEFORE ${PROJECT_BINARY_DIR}/include)
+
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/bin)
+
+if ( ZEEK_PLUGIN_INTERNAL_BUILD )
+    set(ZEEK_SPICY_PLUGIN_INTERNAL_BUILD "yes")
+else ()
+    set(ZEEK_SPICY_PLUGIN_INTERNAL_BUILD "no")
+endif ()
+
+if ( ZEEK_DEBUG_BUILD )
+    set(ZEEK_DEBUG_BUILD "yes") # Prettify output
+else ()
+    set(ZEEK_DEBUG_BUILD "no")
+endif ()
+
+if ( NOT CMAKE_BUILD_TYPE )
+    # We follow Zeek's build mode by default.
+    if ( ZEEK_DEBUG_BUILD )
+        set(CMAKE_BUILD_TYPE "Debug" CACHE STRING "" FORCE)
+    else ()
+        set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "" FORCE)
     endif ()
+endif ()
 
-    # CMake uses -O2 by default with RelWithDebInfo.
-    string(REPLACE "-O2" "-O3" CMAKE_CXX_FLAGS_RELWITHDEBINFO ${CMAKE_CXX_FLAGS_RELWITHDEBINFO})
-    set(SPICY_CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}") # treat as same
+# TODO: We should be able to support JIT with Zeek-internal builds as well now.
+if ( SPICY_HAVE_TOOLCHAIN AND NOT ZEEK_SPICY_PLUGIN_INTERNAL_BUILD )
+    set(SPICY_PLUGIN_USE_JIT "yes")
 else ()
-    set(SPICY_CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}) # save this
-    set(ZEEK_SPICY_PLUGIN_BUILD_DIR ${CMAKE_CURRENT_BINARY_DIR}/plugin)
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${ZEEK_SPICY_PLUGIN_BUILD_DIR}/bin)
+    set(SPICY_PLUGIN_USE_JIT "no")
+endif ()
+
+if ( SPICY_HAVE_TOOLCHAIN )
+    add_subdirectory(src/compiler)
 endif ()
 
-include_directories(BEFORE PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
-include_directories(BEFORE PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/plugin)
-include_directories(BEFORE PRIVATE ${ZEEK_SPICY_PLUGIN_BUILD_DIR}/include)
+###
+### Setup Zeek plugin
+###
+
+list(APPEND CMAKE_MODULE_PATH "${ZEEK_CMAKE_DIR}")
+
+# Prepare environment for Zeek's BroPlugin
+set(BROKER_ROOT_DIR "${ZEEK_PREFIX}")
+set(BRO_CONFIG_CMAKE_DIR "${ZEEK_CMAKE_DIR}")
+set(BRO_CONFIG_INCLUDE_DIR "${ZEEK_INCLUDE_DIRS}")
+set(BRO_CONFIG_PLUGIN_DIR "${ZEEK_PLUGIN_DIR}")
+set(BRO_CONFIG_PREFIX "${ZEEK_PREFIX}")
+set(BRO_PLUGIN_BASE "${PROJECT_SOURCE_DIR}")
+set(BinPAC_ROOT_DIR "${ZEEK_PREFIX}")
+set(CAF_ROOT_DIR "${ZEEK_PREFIX}")
+
+set(save_c_flags ${CMAKE_C_FLAGS})
+set(save_cxx_flags ${CMAKE_CXX_FLAGS})
```

Comment:
Just as a comment, that this is required looks like a bug in `BroPlugin.cmake` to me. Let's maybe open an issue there to make this easier to consume.

Reply:
It's part of a larger task to modernize the Zeek-side plugin infrastructure (but unclear how to best do that so that we don't break things).